samza-core/src/main/java/org/apache/samza/table/remote/TableAsyncHelper.java

```diff
@@ -0,0 +1,167 @@
+            return r;
+          });
+  }
+
+  /**
+   * Execute an async request given a table key
+   * @param key key of the table record
+   * @param method method to be executed
+   * @param timer latency metric to be updated
+   * @param <T> return type
+   * @return CompletableFuture of the operation
+   */
+  public <T> CompletableFuture<T> execute(
+      K key, Function<K, CompletableFuture<T>> method, Timer timer) {
+    final long startNs = System.nanoTime();
+    CompletableFuture<T> ioFuture = rateLimiter.isRateLimited() ?
+        CompletableFuture
+            .runAsync(() -> rateLimiter.throttle(key), tableExecutor)
+            .thenCompose((r) -> method.apply(key)) :
+        method.apply(key);
+    return decorateFuture(ioFuture, startNs, timer);
+  }
+
+  /**
+   * Execute an async request given a table record (key+value)
+   * @param key key of the table record
+   * @param value value of the table record
+   * @param method method to be executed
+   * @param timer latency metric to be updated
+   * @return CompletableFuture of the operation
+   */
+  public CompletableFuture<Void> execute(
+      K key, V value, BiFunction<K, V, CompletableFuture<Void>> method, Timer timer) {
+    final long startNs = System.nanoTime();
+    CompletableFuture<Void> ioFuture = rateLimiter.isRateLimited() ?
+        CompletableFuture
+            .runAsync(() -> rateLimiter.throttle(key, value), tableExecutor)
+            .thenCompose((r) -> method.apply(key, value)) :
+        method.apply(key, value);
+    return decorateFuture(ioFuture, startNs, timer);
+  }
+
+  /**
+   * Execute an async request given a collection of table keys
+   * @param keys collection of keys
+   * @param method method to be executed
+   * @param timer latency metric to be updated
+   * @return CompletableFuture of the operation
+   */
+  public <T> CompletableFuture<T> execute(
+      Collection<K> keys, Function<Collection<K>, CompletableFuture<T>> method, Timer timer) {
+    final long startNs = System.nanoTime();
+    CompletableFuture<T> ioFuture = rateLimiter.isRateLimited() ?
+        CompletableFuture
+            .runAsync(() -> rateLimiter.throttle(keys), tableExecutor)
+            .thenCompose((r) -> method.apply(keys)) :
+        method.apply(keys);
+    return decorateFuture(ioFuture, startNs, timer);
+  }
+
+  /**
+   * Execute an async request given a collection of table records
+   * @param records list of records
+   * @param method method to be executed
+   * @param timer latency metric to be updated
+   * @return CompletableFuture of the operation
+   */
+  public CompletableFuture<Void> executeRecords(Collection<Entry<K, V>> records,
+      Function<Collection<Entry<K, V>>, CompletableFuture<Void>> method, Timer timer) {
+    final long startNs = System.nanoTime();
+    CompletableFuture<Void> ioFuture = rateLimiter.isRateLimited() ?
+        CompletableFuture
+            .runAsync(() -> rateLimiter.throttleRecords(records), tableExecutor)
```

Comment:
As a separate design choice, can you comment on why we need a separate executor for the throttler? 

IMO, if there are no tokens available, getAsync(), putAsync() should block after some back-off mechanism. Otherwise, it results in back-pressure that fills up system memory. 

Additionally, we should explore if it makes sense to use the same thread-pool instead of creating a separate one for the throttler